cmd/snap/cmd_run.go

```diff
@@ -216,6 +220,68 @@ func snapRunHook(snapName, snapRevision, hookName string) error {
 	return runSnapConfine(info, hook.SecurityTag(), snapName, "", hook.Name, nil)
 }
 
+func migrateXauthority(info *snap.Info) error {
+	u, err := userCurrent()
+	if err != nil {
+		return fmt.Errorf(i18n.G("cannot get the current user: %v"), err)
+	}
+
+	xauthPath := osGetenv("XAUTHORITY")
+	if len(xauthPath) == 0 || !osutil.FileExists(xauthPath) {
+		// Nothing to do for us. Most likely running outside of any
+		// graphical X11 session.
+		return nil
+	}
```

Comment:
In thinking about this more, there is an attack I'd like to add a little more protection against. Consider:

1. sysadmin updates sudoers to allow running 'sudo /snap/bin/foo' by some non-admin user
2. that user does `XAUTHORITY=/etc/shadow sudo /snap/bin/foo`
3. /etc/shadow would be copied to /run/user/0/Xauthority
4. foo is somehow controllable by the non-admin user (eg, flaw in foo, developer of foo, etc) such that /run/user/0/Xauthority (ie, /etc/shadow) can be shipped off, copied to SNAP_DATA, etc

This sort of attack is precisely why I recommended validating the xauth file (ie, ValidateXauthority, below) but I think a nice hardening/defensive-coding measure would be to check if XAUTHORITY is in /tmp to help guard against if our validation code has a bug. The whole reason we are doing this is because /tmp is in a different mount namespace and some distros put it in there, setting XAUTHORITY. Let's just fix that rather than trying to be overly generic. As such, please, after you've stored off xauthPath (ie, right where this comment is), please add something to the effect of:
```
	// Abs() also calls Clean()
	// https://golang.org/pkg/path/filepath/#Abs
	xauthPathAbs, err := filepath.Abs(xauthPath)
	if err != nil {
		return nil
	}
	// remove all symlinks from path
	xauthPathCan, err := filepath.EvalSymlinks(xauthPathAbs)
	if err != nil {
		return nil
	}
	// verify XAUTHORITY matches the cleaned path
	if xauthPath != xauthPathCan {
		logger.Noticef("WARNING: %s != %s\n", xauthPath, xauthPathCan)
		return nil
	}
	// Only do the migration from /tmp since /tmp is what is in a different mount
	// namespace
	if !strings.HasPrefix(xauthPath, "/tmp") {
		logger.Noticef("WARNING: %s is not in /tmp", xauthPath)
		return nil
	}
```